python-checks/src/main/java/org/sonar/python/checks/IncompatibleOperandsCheck.java

```diff
@@ -0,0 +1,163 @@
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("+", new SpecialMethodNames("__add__", "__radd__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("-", new SpecialMethodNames("__sub__", "__rsub__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("*", new SpecialMethodNames("__mul__", "__rmul__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("@", new SpecialMethodNames("__matmul__", "__rmatmul__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("/", new SpecialMethodNames("__truediv__", "__rtruediv__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("//", new SpecialMethodNames("__floordiv__", "__rfloordiv__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("%", new SpecialMethodNames("__mod__", "__rmod__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("**", new SpecialMethodNames("__pow__", "__rpow__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("<<", new SpecialMethodNames("__lshift__", "__rlshift__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put(">>", new SpecialMethodNames("__rshift__", "__rrshift__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("&", new SpecialMethodNames("__and__", "__rand__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("^", new SpecialMethodNames("__xor__", "__rxor__"));
+    BINARY_SPECIAL_METHODS_BY_OPERATOR.put("|", new SpecialMethodNames("__or__", "__ror__"));
+  }
+
+  // https://github.com/python/mypy/blob/e97377c454a1d5c019e9c56871d5f229db6b47b2/mypy/semanal_classprop.py#L16-L46
+  private static final Set<Set<String>> HARDCODED_COMPATIBLE_TYPES = new HashSet<>();
+  static {
+    HARDCODED_COMPATIBLE_TYPES.add(new HashSet<>(Arrays.asList(INT, FLOAT, COMPLEX)));
+    HARDCODED_COMPATIBLE_TYPES.add(new HashSet<>(Arrays.asList(STR, "unicode", "bytearray", "bytes", "memoryview")));
+  }
+
+  private static class SpecialMethodNames {
+    private final String left;
+    private final String right;
+
+    public SpecialMethodNames(String left, String right) {
+      this.left = left;
+      this.right = right;
+    }
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpression binaryExpression) {
+    InferredType leftType = binaryExpression.leftOperand().type();
+    InferredType rightType = binaryExpression.rightOperand().type();
+    checkComparisonOperands(binaryExpression.operator(), leftType, rightType);
+    super.visitBinaryExpression(binaryExpression);
+  }
+
+  @Override
+  public void visitUnaryExpression(UnaryExpression unaryExpression) {
+    InferredType type = unaryExpression.expression().type();
+    Token operator = unaryExpression.operator();
+    String memberName = UNARY_SPECIAL_METHODS_BY_OPERATOR.get(operator.value());
+    if (memberName != null && !type.canHaveMember(memberName)) {
+      addIssue(operator, "Fix this invalid \"" + operator.value() + "\" operation on a type which doesn't support it.");
+    }
+    super.visitUnaryExpression(unaryExpression);
+  }
+
+  private void checkComparisonOperands(Token operator, InferredType left, InferredType right) {
```

Comment:
This method is used for more than comparison operations. It might be better to have more generic names to reflect that.